medusa/search/core.py

```diff
@@ -332,32 +329,60 @@ def pick_best_result(results):  # pylint: disable=too-many-branches
-        if Quality.is_higher_quality(best_result.quality, cur_result.quality, allowed_qualities, preferred_qualities):
-            best_result = cur_result
-        elif best_result.quality == cur_result.quality:
-            if any(ext in cur_result.name.lower() for ext in preferred_words):
-                log.info(u'Preferring {0} (preferred words)', cur_result.name)
-                best_result = cur_result
-            if cur_result.proper_tags:
-                log.info(u'Preferring {0} (repack/proper/real/rerip over nuked)', cur_result.name)
-                best_result = cur_result
-            if any(ext in best_result.name.lower() for ext in undesired_words) and not any(ext in cur_result.name.lower() for ext in undesired_words):
-                log.info(u'Unwanted release {0} (contains undesired word(s))', cur_result.name)
-                best_result = cur_result
-
-    if best_result:
-        log.debug(u'Picked {0} as the best', best_result.name)
+        wanted_results.append(cur_result)
+
+    if wanted_results:
+        log.debug(u'Found wanted results.')
     else:
-        log.debug(u'No result picked.')
+        log.debug(u'No wanted results found.')
+
+    return wanted_results
+
+
+def pick_result(results):
+    """Pick a result out of a list of wanted candidates."""
+    if not results:
+        log.debug(u'No results to pick from.')
+        return
+
+    log.debug(u'Picking the best result out of {0}', [x.name for x in results])
+
+    wanted_results = {}
+    sorted_results = []
+
+    preferred_words = []
+    if app.PREFERRED_WORDS:
+        preferred_words = [_.lower() for _ in app.PREFERRED_WORDS]
+    undesired_words = []
+    if app.UNDESIRED_WORDS:
+        undesired_words = [_.lower() for _ in app.UNDESIRED_WORDS]
+
+    for result in results:
+        wanted_results[result] = 0
+
+        if any(word in result.name.lower() for word in undesired_words):
+            log.info(u'Penalizing release {0} (contains undesired word(s))', result.name)
+            wanted_results[result] -= 2
+
+        if any(word in result.name.lower() for word in preferred_words):
+            log.info(u'Rewarding release {0} (preferred word(s))', result.name)
+            wanted_results[result] += 2
+
+        if sorted_results:
+            allowed_qualities, preferred_qualities = result.series.current_qualities
+            if Quality.is_higher_quality(sorted_results[0][0].quality, result.quality,
+                                         allowed_qualities, preferred_qualities):
+                log.info(u'Rewarding release {0} (higher quality)', result.name)
+                wanted_results[result] += 1
+
+            elif result.proper_tags and sorted_results[0][0].quality == result.quality:
+                log.info(u'Rewarding release {0} (repack/proper/real/rerip)', result.name)
+                wanted_results[result] += 1
+
+        sorted_results = sorted(iteritems(wanted_results), key=operator.itemgetter(1), reverse=True)
```

Comment:
1. I don't really like this solution, it seems a bit excessive to sort the results with each loop iteration.
	I'm not sure sorting is needed - https://stackoverflow.com/a/13145419/7597273
	```py
	# max() returns the tuple with the highest score (result, score), then we get the first item
	best_result = max(iteritems(wanted_results), key=operator.itemgetter(1))[0]
	```
	My opinion is to go back to using `best_result` instead of `sorted_results`.
	I think it's cleaner and more effective.

2. What if two or more results have the same score?

3. Just throwing it out there: we may want to list the scores?

Reply:
1. You are right, it's not really needed, but I also don't think it's that expensive. The idea behind it was to show the results with scores at the end of the iteration (as you suggested in 3.). If we don't want that it can be removed of course. IMO it's just better to have more information vs a few ms faster code.

2. In that case, the first result with the same score will be picked.

3. See 1. 😄

Reply:
> The idea behind it was to show the results with scores at the end of the iteration (as you suggested in 3.)

Reply:
> The idea behind it was to show the results with scores at the end of the iteration (as you suggested in 3.)
> IMO it's just better to have more information vs a few ms faster code.

I still think sorting the results after every result is a bit much.
And I don't see how it gives you more information?
You can sort `wanted_results` once before printing the table to the log.

I don't have any concrete data but it makes sense that sorting is heavier on memory and speed.

Reply:
>My opinion is to go back to using best_result instead of sorted_results.
I think it's cleaner and more effective.
>...
>I still think sorting the results after every result is a bit much.

Just an indirectly related  side-note here, regarding optimizing search efficiency, I think that all the "get" operations inside the results-loop related to fetching 
 ignored/preferred/undesired/required "words" can be moved to before that loop:

from:

  ignored_words = series_obj.show_words().ignored_words
  required_words = series_obj.show_words().required_words
...
        preferred_words = []
        if app.PREFERRED_WORDS:
            preferred_words = [_.lower() for _ in app.PREFERRED_WORDS]
        undesired_words = []
        if app.UNDESIRED_WORDS:
            undesired_words = [_.lower() for _ in app.UNDESIRED_WORDS]

Can be moved to before the loop--->

ignored_words = results[0].series.show_words().ignored_words
required_words = results[0].series.show_words().required_words
etc...

I'm sure this will save a few milliseconds too... Just my 2 cents....

Reply:
I remembered why the sorting after every iteration is needed: it's to allow this part of the code to work correctly: https://github.com/pymedusa/Medusa/blob/07fc0dff32083b99c2b58bbe1be30849e52bcc72/medusa/search/core.py#L372-L380 
We need to compare to the current best result, so either we use max() after every iteration (loosing the sorted list, or requiring an additional list) or we just keep it as it is.

Reply:
@rafi-d 
Thank you, but:
1. `preferred_words` and `undesired_words` are not in loops.
2. I think there's always a chance a result with a different series was returned from the providers, so I don't think we can do that.

@medariox 
> I remembered why the sorting after every iteration is needed: it's to allow this part of the code to work correctly:

That's exactly my point, though. It's not **needed**.
If not for efficiency, then for readability:
`Quality.is_higher_quality(best_result.quality, result.quality` **vs.** `Quality.is_higher_quality(sorted_results[0][0].quality, result.quality`

> We need to compare to the current best result, so either we use max() after every iteration (loosing the sorted list, or requiring an additional list) or we just keep it as it is.

With the `max` solution you'd lose the sorted list, yes, but why would it require an additional list?

Reply:
There's also another solution I can now think of: `OrderedDict`.
You can sort the wanted results every loop iteration and it would probably be more efficient than both the `sorted` and `max` solutions.

Reply:
@sharkykh 
>2.I think there's always a chance a result with a different series was returned from the providers, >so I don't think we can do that.
Well, theoretically, yes, it's possible, I can only say that for the last couple of months that I am using this change here, everything looks fine, nothing strange has happened... :) 
End result could have been downloading a wrong series, either way... 

Anyways, global words are the same, and the only difference is the per-series words, so I guess you can check it out of the loop in cases were there aren't any local words defined...

Reply:
If it's theoretically possible, then it's also theoretically possible your last couple of months of testing, your enabled providers, your specific search cases, and your ignored/required/preferred/undesired words have all avoided this theoretical scenario.

Looks like this is the code that filters results that are not wanted series / episodes:
https://github.com/pymedusa/Medusa/blob/07fc0dff32083b99c2b58bbe1be30849e52bcc72/medusa/search/core.py#L283-L290

It's executed before filtering the result based on the words. If the first result is not the series you were searching for then it would could get the wrong data.

> Also, you might pass the series as parameter to pick_best_result()...